### PR TITLE
Upstream Helm chart adapted to our needs

### PR DESCRIPTION
**Pull request by UnderGreen**
### Goals
Have working stolon with PostgreSQL 12, which was impossible without picking up stolon from the upstream repository.

* Post-install hook is no implemented as stolonctl needs reworking. Stolonctl depends on the upstream stolon code and that code was refactored, which needs to be mirrored in stolonctl code. But this I would like to do in another PR.
* No upgrade is implemented as it should be in another PR, which could be big enough by itself.
* Chart path on disk was moved into resources directly as by the gravity packaging concept content on the resources directory is the application/chart. So, I made it more consistent with the concept.
* Etcd backend is replaced with Kubernetes backend(configmap under the hood) for the sake of simplicity to change the configuration on working cluster.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -123,23 +123,23 @@ $(TARBALL):
 	$(GRAVITY) package export $(REPOSITORY)/$(NAME):$(VERSION) $(TARBALL) $(EXTRA_GRAVITY_OPTIONS)
 
 .PHONY: import
-import: images $(BUILD_DIR)/resources/app.yaml
+import: images
 	sed -i "s#gravitational.io/cluster-ssl-app:0.0.0+latest#gravitational.io/cluster-ssl-app:$(CLUSTER_SSL_APP_VERSION)#" resources/app.yaml
-	sed -i "s/tag: latest/tag: $(VERSION)/g" resources/charts/stolon/values.yaml
-	sed -i "s/0.1.0/$(VERSION)/g" resources/charts/stolon/Chart.yaml
+	sed -i "s/tag: latest/tag: $(VERSION)/g" resources/values.yaml
```
Comment (a-palchikov) [inline Makefile]: `sed` will break if this is built from darwin. I'd suggest to use a macro to make it distro-independent:
```
define replace
	sed -i "$1" $2
endef

ifeq ($(OS),darwin)
define replace
	sed -i '' "$1" $2
endef
endif
```
and then use it:
```
$(call replace,"s/tag: $(VERSION)/tag: latest/g",resources/values.yaml)
```

File: images/stolon/Dockerfile [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
     cd $REPO_PATH && git checkout $STOLON_REVISION
 
 WORKDIR $REPO_PATH
 
-RUN mkdir -p bin && \
-    CGO_ENABLED=0 go build -a -installsuffix cgo -ldflags '-s' -o bin/stolon-sentinel ${REPO_PATH}/cmd/sentinel && \
-    CGO_ENABLED=0 go build -a -installsuffix cgo -ldflags '-s' -o bin/stolon-proxy ${REPO_PATH}/cmd/proxy && \
-    CGO_ENABLED=0 go build -a -installsuffix cgo -ldflags '-s' -o bin/stolontool ${REPO_PATH}/cmd/stolonctl && \
-    go build -a -installsuffix cgo -ldflags '-s' -o bin/stolon-keeper ${REPO_PATH}/cmd/keeper
+RUN make
```
Comment (a-palchikov) [inline images/stolon/Dockerfile]: why make `make` a separate layer? This can be done as part of the previous RUN

File: images/stolon/Dockerfile [hunk 3]
```diff
@@ -1,22 +1,18 @@
-ARG PGVERSION=9.6
+ARG PGVERSION=12
 # base build image
-FROM quay.io/gravitational/debian-venti:go1.13.8-buster AS builder
-ARG STOLON_REVISION=6353dbc5542d7243bd2bd5256f0a945fdc1f5c23
+FROM quay.io/gravitational/debian-venti:go1.16.3-buster AS builder
+ARG STOLON_REVISION=v0.16.0
```
Comment (a-palchikov) [inline images/stolon/Dockerfile]: Also, would be nice to move all configuration outside of the dockerfiles - i.e. be able to override versions from a makefile - not sure this is the case now.

File: resources/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
+# Stolon Helm Chart
+
+* Installs [Stolon](https://github.com/sorintlab/stolon) (HA PostgreSQL cluster)
+* Based on [lwolf/stolon-chart](https://github.com/lwolf/stolon-chart) and [stolon examples](https://github.com/sorintlab/stolon/tree/master/examples/kubernetes/statefulset)
+
+
+Official upstream `helm/charts` support was deprecated on Nov 13, 2020. This a
+community effor to maintain the helm chart. Please create an issue if you're
```
Comment (a-palchikov) [inline resources/README.md:8]:
```suggestion
community effort to maintain the helm chart. Please create an issue if you're
```

File: resources/README.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+interested on helping to maintain this chart.
+
+
+## TL;DR;
+
+```console
+$ helm repo add iomedstolon https://iomedhealth.github.io/helm_stolon/
```
Comment (a-palchikov) [inline resources/README.md]: Is the repo correct?

File: resources/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+To install the chart with the release name `my-release`:
+
+```console
+$ helm install --name my-release iomedstolon/stolon
+```
+
+## Backend
+
+Kubernetes is the default store backend. `consul`, `etcdv2` or `etcdv3` can also be used as the store backend.
```
Comment (a-palchikov) [inline resources/README.md]: Kubernetes should be `"kubernetes"` e.g. in quotes since it should be a backend name?

File: resources/app.yaml [hunk 7]
```diff
@@ -106,10 +106,10 @@ nodeProfiles:
 hooks:
   install:
     job: file://install.yaml
-  postInstall:
-    job: file://postInstall.yaml
-  preUpdate:
-    job: file://preUpdateHook.yaml
+  #postInstall:
```
Comment (a-palchikov) [inline resources/app.yaml:109]: remove these?

File: resources/install.sh [hunk 8]
```diff
@@ -9,9 +9,8 @@ then
 fi
 
 set +e
```
Comment (a-palchikov) [inline resources/install.sh]: set +eu

File: resources/templates/rolebinding.yaml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 {{- if .Values.rbac.create -}}
-apiVersion: rbac.authorization.k8s.io/v1
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (a-palchikov) [inline resources/templates/rolebinding.yaml]: can you add some context here - why is the version changed?

File: resources/values.yaml [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+replicationUsername: "repluser"
+## password for the replication user
+replicationPassword:
+replicationForcePassword: false
+
+## backend could be one of the following: consul, etcdv2, etcdv3 or kubernetes
+store:
+  backend: kubernetes
+#  endpoints: "http://stolon-consul:8500"
```
Comment (a-palchikov) [inline resources/values.yaml]: remove or does it hint about the default?

**Review by a-palchikov — COMMENTED** [review]
made an initial pass but did look into templates in detail though

File: images/stolon/Dockerfile (hunk 3 above)
Comment (UnderGreen) [reply images/stolon/Dockerfile]: ARG could be used outside of docker files with the `--arg` parameter. So, yes, I will move the values into Makefile.

File: images/stolon/Dockerfile (hunk 2 above)
Comment (UnderGreen) [reply images/stolon/Dockerfile]: The previous run is before the `WORKDIR` command which is important for the make as we need to change working directory prior to using make.

File: resources/app.yaml (hunk 7 above)
Comment (UnderGreen) [reply resources/app.yaml:109]: They are intended to be reimplemented in another PR

File: Makefile (hunk 1 above)
Comment (UnderGreen) [reply Makefile]: Is tele binary available for the Darwin architecture? I am not aware of it then. :(

Comment (a-palchikov) [reply Makefile]: `tele` was (almost) always available for darwin. But even `gravity` works on darwin (with some minor changes)

File: images/stolon/Dockerfile (hunk 2 above)
Comment (a-palchikov) [reply images/stolon/Dockerfile]: I understand. `WORKDIR` could be moved before the previous RUN.
Also, it's not necessary to `cd` as you could condense everything into a single`git clone`:
```
WORKDIR $REPO_PATH
RUN set -eux; \
    apt-get update && \
    apt-get install -qq -y git && \
    git clone https://github.com/sorintlab/stolon.git $REPO_PATH --depth=1 --branch $STOLON_REVISION && \
    make

```

File: Makefile (hunk 1 above)
Comment (UnderGreen) [reply Makefile]:
```
[ERROR]: tele build is not supported on darwin, only Linux is supported
```

Comment (a-palchikov) [reply Makefile]: Yes, I know but it's trivial to remove an otherwise it should work. Sorry for confusion.

**Pull request closed without merging**